### PR TITLE
fix(core): MediaObserver can report 1..n activations

### PR DESCRIPTION
**Pull request by ThomasBurleson**
Previous versions of MediaObserver suffered from a significant design-flaw. 

Those versions assumed that a breakpoint change would only activate/match a single mediaQuery. Additionally those versions would not (by default) report overlapping activations. Applications interested in notifications for all current activations would therefore not receive proper event-notifications.

The current enhancements provide several features:

* Report 1..n mediaQuery activations (matches == true) in a single event
* Report activations sorted by descending priority
* By default, reports include overlapping breakpoint activations
* Debounce notifications to a single grouped event
  > useful to reduce browser reflow thrashing

BREAKING CHANGE:

The stream data type is now **MediaChange[]** instead of *MediaChange* and`media$` is deprecated in favor of `asObservable()`. 

* `filterOverlaps` now defaults to `false`
* `media$` is now a `Observable< MediaChange[] >` instead of `Observable< MediaChange >`

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@CaerusKaru - ready for review.

File: src/apps/demo-app/src/app/media-query-status/media-query-status.component.ts [hunk 1]
```diff
@@ -1,25 +1,25 @@
-import {ChangeDetectionStrategy, Component} from '@angular/core';
+import {Component} from '@angular/core';
 import {MediaChange, MediaObserver} from '@angular/flex-layout';
-import {Observable} from 'rxjs';
+import {Observable} from 'rxjs/internal/Observable';
```
Comment (CaerusKaru) [inline src/apps/demo-app/src/app/media-query-status/media-query-status.component.ts]: Just `rxjs`?

File: src/lib/core/match-media/match-media.spec.ts [hunk 2]
```diff
@@ -141,46 +145,47 @@ describe('match-media', () => {
...
+    it('reports mediaQuery de-activations', () => {
+      const lookupMediaQuery = (alias: string) => {
+        const bp: BreakPoint = breakPoints.findByAlias(alias) as BreakPoint;
+        return bp.mediaQuery;
+      };
+      let activationCount = 0, deactivationCount = 0;
+      let subscription = watchMedia('', (change: MediaChange) => {
         if (change.matches) {
-          ++activationCount;
+          activationCount += 1;
```
Comment (CaerusKaru) [inline src/lib/core/match-media/match-media.spec.ts]: This isn't exactly a sticking point, but why not `activationPoint++`?

File: src/lib/core/media-observer/media-observer.ts [hunk 3]
```diff
@@ -43,41 +41,67 @@ import {BreakPointRegistry, OptionalBreakPoint} from '../breakpoints/break-point
...
  *    }
  *  }
  */
 @Injectable({providedIn: 'root'})
 export class MediaObserver {
+
   /**
-   * Whether to announce gt-<xxx> breakpoint activations
+   * @deprecated Use `asObservable()` instead.
+   * @deletion-target v7.0.0-beta.24
```
Comment (CaerusKaru) [inline src/lib/core/media-observer/media-observer.ts]: deletion-target should be +2 releases which would put this at beta.26 (assuming it's released with beta 24)

File: src/lib/core/media-observer/media-observer.ts [hunk 4]
```diff
@@ -43,41 +41,67 @@ import {BreakPointRegistry, OptionalBreakPoint} from '../breakpoints/break-point
...
+   * @breaking-change 7.0.0-beta.24
```
Comment (CaerusKaru) [inline src/lib/core/media-observer/media-observer.ts]: Unneeded

File: tslint.json [hunk 5]
```diff
@@ -25,7 +25,7 @@
     "no-trailing-whitespace": true,
     "no-bitwise": true,
     "no-shadowed-variable": true,
-    "no-unused-expression": true,
+    "no-unused-expression": [true, "allow-fast-null-checks"],
```
Comment (CaerusKaru) [inline tslint.json]: Why is this here?

File: src/lib/core/media-observer/media-observer.ts [hunk 6]
```diff
@@ -93,53 +117,79 @@ export class MediaObserver {
...
+/**
+ * Find associated breakpoint (if any)
+ */
+function toMediaQuery(query: string, locator: BreakPointRegistry) {
+  const bp = locator.findByAlias(query) || locator.findByQuery(query);
+  return bp ? bp.mediaQuery : query;
+}
+
+/** HOF to sort the breakpoints by priority */
+export function sortChangesByPriority(a: MediaChange, b: MediaChange): number {
```
Comment (CaerusKaru) [inline src/lib/core/media-observer/media-observer.ts]: Isn't this exported somewhere else?

File: src/lib/core/media-observer/media-observer.ts [hunk 7]
```diff
@@ -43,41 +41,67 @@ import {BreakPointRegistry, OptionalBreakPoint} from '../breakpoints/break-point
...
   /**
-   * Whether to announce gt-<xxx> breakpoint activations
+   * @deprecated Use `asObservable()` instead.
+   * @deletion-target v7.0.0-beta.24
+   * @breaking-change 7.0.0-beta.24
    */
-  filterOverlaps = true;
-  readonly media$: Observable<MediaChange>;
+  get media$(): Observable<MediaChange[]> {
+    return this._media$;
```
Comment (CaerusKaru) [inline src/lib/core/media-observer/media-observer.ts]: Note: we need to support the old behavior here. Maybe add `deprecatedMedia$`?

File: src/lib/core/media-observer/media-observer.ts (hunk 3 above)
Comment (ThomasBurleson) [reply src/lib/core/media-observer/media-observer.ts]: deprecation for 24
deletion for 25

File: src/lib/core/media-observer/media-observer.ts (hunk 4 above)
Comment (ThomasBurleson) [reply src/lib/core/media-observer/media-observer.ts]: deprecation for 24
deletion for 25

File: src/lib/core/media-observer/media-observer.ts (hunk 7 above)
Comment (ThomasBurleson) [reply src/lib/core/media-observer/media-observer.ts]: fixed.

File: src/lib/core/media-observer/media-observer.ts (hunk 3 above)
Comment (CaerusKaru) [reply src/lib/core/media-observer/media-observer.ts]: Compromise accepted 😄

File: src/lib/core/media-observer/media-observer.ts (hunk 6 above)
Comment (ThomasBurleson) [reply src/lib/core/media-observer/media-observer.ts]: Different data types.

**Comment by ThomasBurleson** [comment]
*  `media$: Observable< MediaChange >` behavior restored. 
*  `asObservable(): Observable< MediaChange[] >`

File: src/lib/core/media-observer/media-observer.ts (hunk 6 above)
Comment (CaerusKaru) [reply src/lib/core/media-observer/media-observer.ts]: Can we make the original a generic?
